src/storage/src/hummock/store/version.rs

```diff
@@ -154,48 +170,33 @@ impl StagingVersion {
         max_epoch_inclusive: HummockEpoch,
         table_id: TableId,
         table_key_range: &'a TableKeyRange,
-    ) -> (
-        impl Iterator<Item = &ImmutableMemtable> + 'a,
-        impl Iterator<Item = &SstableInfo> + 'a,
-    ) {
+    ) -> impl Iterator<Item = &StagingData> + 'a {
         let (ref left, ref right) = table_key_range;
         let left = left.as_ref().map(|key| TableKey(key.0.as_ref()));
         let right = right.as_ref().map(|key| TableKey(key.0.as_ref()));
-        let overlapped_imms = self.imm.iter().filter(move |imm| {
+        // reverse to make the newer data come first.
+        self.data.iter().rev().filter(move |data| {
             // retain imm which is overlapped with (min_epoch_exclusive, max_epoch_inclusive]
-            imm.min_epoch() <= max_epoch_inclusive
-                && imm.table_id == table_id
-                && range_overlap(
-                    &(left, right),
-                    &imm.start_table_key(),
-                    Included(&imm.end_table_key()),
-                )
-        });
-
-        // TODO: Remove duplicate sst based on sst id
-        let overlapped_ssts = self
-            .sst
-            .iter()
-            .filter(move |staging_sst| {
-                let sst_max_epoch = *staging_sst.epochs.last().expect("epochs not empty");
-                sst_max_epoch <= max_epoch_inclusive
-            })
-            .flat_map(move |staging_sst| {
-                // TODO: sstable info should be concat-able after each streaming table owns a read
-                // version. May use concat sstable iter instead in some cases.
-                staging_sst
-                    .sstable_infos
-                    .iter()
-                    .map(|sstable| &sstable.sst_info)
-                    .filter(move |sstable: &&SstableInfo| {
-                        filter_single_sst(sstable, table_id, table_key_range)
```

Comment:
It is somehow strange that `prune_overlap` uses key range + epoch to filter out imm but uses only epoch to filter out staging sst after this PR. Why do we move `filter_single_sst` to `iter` and `get`? I think we should keep the pruning rule consistent for imm and staging sst. That is either we prune key range + epoch for both imm and staging sst, or we only prune epoch for both imm and staging sst. I think the reason why we prune both here prior to this PR is because we want to reduce the number of clone after `prune_overlap` in `read_filter_for_version`.

Reply:
No. This PR will call `prune_overlaping_level` for staging data before really query sstable.  It just move this condition to `iter_inner`

Reply:
I know. Correctness is still guaranteed in this PR. This is more like a style issue.